cpp/include/raft/neighbors/detail/ann_mg.cuh

```diff
@@ -0,0 +1,645 @@
+      index_.emplace(std::move(ivf_flat::deserialize<T, IdxT>(handle, is)));
+    } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+      index_.emplace(std::move(ivf_pq::deserialize<T, IdxT>(handle, is)));
+    } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+      index_.emplace(std::move(cagra::deserialize<T, IdxT>(handle, is)));
+    }
+  }
+
+  const IdxT size() const
+  {
+    if constexpr (std::is_same<AnnIndexType, ivf_flat::index<T, IdxT>>::value) {
+      return index_.value().size();
+    } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+      return index_.value().size();
+    } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+      return index_.value().size();
+    }
+  }
+
+ private:
+  std::optional<AnnIndexType> index_;
+};
+
+template <typename AnnIndexType, typename T, typename IdxT>
+class ann_mg_index {
+ public:
+  ann_mg_index(const std::vector<int>& dev_list, dist_mode mode = SHARDING)
+    : mode_(mode),
+      root_rank_(0),
+      num_ranks_(dev_list.size()),
+      dev_ids_(dev_list),
+      nccl_comms_(dev_list.size())
+  {
+    init_device_resources();
+    init_nccl_clique();
+  }
+
+  // deserialization
+  ann_mg_index(const raft::resources& handle,
+               const std::string& filename) {
+      std::ifstream is(filename, std::ios::in | std::ios::binary);
+      if (!is) { RAFT_FAIL("Cannot open file %s", filename.c_str()); }
+
+      mode_ = (raft::neighbors::mg::dist_mode)deserialize_scalar<int>(handle, is);
+      root_rank_ = 0;
+      num_ranks_ = deserialize_scalar<int>(handle, is);
+      dev_ids_.resize(num_ranks_);
+      std::iota(std::begin(dev_ids_), std::end(dev_ids_), 0);
+      nccl_comms_.resize(num_ranks_);
+
+      init_device_resources();
+      init_nccl_clique();
+
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+        auto& ann_if = ann_interfaces_.emplace_back();
+        ann_if.deserialize(dev_resources_[rank], is);
+      }
+
+      is.close();
+  }
+
+  ann_mg_index(const ann_mg_index&)                    = delete;
+  ann_mg_index(ann_mg_index&&)                         = default;
+  auto operator=(const ann_mg_index&) -> ann_mg_index& = delete;
+  auto operator=(ann_mg_index&&) -> ann_mg_index&      = default;
+
+  void init_device_resources() {
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      dev_resources_.emplace_back();
+    }
+  }
+
+  void init_nccl_clique() {
+    RAFT_NCCL_TRY(ncclCommInitAll(nccl_comms_.data(), num_ranks_, dev_ids_.data()));
+    for (int rank = 0; rank < num_ranks_; rank++) {
+      RAFT_CUDA_TRY(cudaSetDevice(dev_ids_[rank]));
+      raft::comms::build_comms_nccl_only(&dev_resources_[rank], nccl_comms_[rank], num_ranks_, rank);
```

Comment:
The NCCL initialization seems to be "one process multiple GPUs". 

Is it possible to adapt it to "one process or thread one GPU"? May have to use something like std::thread. But the benefit is to enable the APIs of the PR to be reusable to Dask/Spark. Both Dask and Spark currently comply with one process one GPU when initializing NCCL.

Reply:
The single process solution was better suited to implement the much requested feature in RAFT for now. But, I agree that in the end we should definitely look into making it possible to make things run on Dask/Spark. This would probably involve the use of multiple processes/threads and a much broader use of NCCL.
cc @cjnolet